rust/tw_hd_wallet/src/ton/mod.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright © 2017 Trust Wallet.
+
+use crate::bip39::normalize_mnemonic;
+use crate::ton::mnemonic::validate_mnemonic_words;
+use crate::{WalletError, WalletResult};
+use tw_hash::hmac::hmac_sha512;
+use tw_hash::pbkdf2::pbkdf2_hmac_sha512;
+use tw_hash::{H256, H512};
+use tw_keypair::ed25519::sha512::KeyPair;
+use zeroize::ZeroizeOnDrop;
+
+pub const TON_WALLET_SALT: &[u8] = b"TON default seed";
+pub const TON_WALLET_PBKDF_ITERATIONS: u32 = 100_000;
+
+/// Used to verify a pair of mnemonic and passphrase only.
+const TON_BASIC_SEED_SALT: &[u8] = b"TON seed version";
+/// Equals to `(TON_WALLET_PBKDF_ITERATIONS as f64 / 256.0).floor() as u32`.
+const TON_BASIC_SEED_ROUNDS: u32 = 390;
+
+/// Used to verify a pair of mnemonic and passphrase only.
+const TON_PASSPHRASE_SEED_SALT: &[u8] = b"TON fast seed version";
+const TON_PASSPHRASE_SEED_ROUNDS: u32 = 1;
+
+mod mnemonic;
+
+#[derive(Debug, ZeroizeOnDrop)]
+pub struct TonWallet {
+    mnemonic: String,
+    passphrase: Option<String>,
+    entropy: H512,
+    seed: H512,
+}
+
+impl TonWallet {
+    pub const MNEMONIC_WORDS: usize = 24;
+
+    /// Creates `TonWallet` while validating if there should or shouldn't be a passphrase.
+    /// https://github.com/toncenter/tonweb-mnemonic/blob/a338a00d4ca0ed833431e0e49e4cfad766ac713c/src/functions/validate-mnemonic.ts#L20-L28
+    pub fn new(mnemonic: &str, passphrase: Option<String>) -> WalletResult<TonWallet> {
+        let mnemonic = normalize_mnemonic(mnemonic);
+        validate_mnemonic_words(&mnemonic)?;
+
+        let entropy = Self::ton_mnemonic_to_entropy(&mnemonic, passphrase.as_deref());
```

Comment:
Should we validate the passphrase and mnemonic before creating `entropy` to avoid unnecessary computation? (meaning move this after `match` case)

Reply:
Basically no, I'll move it below the passphrase check